File: hash/java/src/main/java/io/github/wmetcalf/rosettasquint/hash/internal/PilMedianFilter.java
package io.github.wmetcalf.rosettasquint.hash.internal;

import java.util.Arrays;

/**
 * Pillow-compatible MedianFilter with size=3 (3×3 windowed median).
 *
 * Matches PIL.ImageFilter.MedianFilter(size=3) exactly:
 * - For each pixel, collect the 9 values in the 3×3 window, clamping (edge-replicating)
 *   at borders so e.g. pixel (0,0) contributes 4 times.
 * - Sort the 9 values, return the value at index 4 (the median of 9).
 *
 * Reference: PIL spec §3.2 (parity hazard 3.2).
 */
public final class PilMedianFilter {

    /**
     * Apply PIL MedianFilter(size=3) to a uint8 grayscale image.
     *
     * @param src  row-major uint8 array [H][W], values 0-255
     * @return new [H][W] array with median filter applied
     */
    public static int[][] apply(int[][] src) {
        int H = src.length;
        int W = src[0].length;
        int[][] out = new int[H][W];
        int[] window = new int[9];

        for (int y = 0; y < H; y++) {
            for (int x = 0; x < W; x++) {
                int idx = 0;
                for (int dy = -1; dy <= 1; dy++) {
                    int sy = clamp(y + dy, 0, H - 1);
                    for (int dx = -1; dx <= 1; dx++) {
                        int sx = clamp(x + dx, 0, W - 1);
                        window[idx++] = src[sy][sx];
                    }
                }
                Arrays.sort(window);
                out[y][x] = window[4];
            }
        }
        return out;
    }

    private static int clamp(int v, int lo, int hi) {
        return v < lo ? lo : (v > hi ? hi : v);
    }

    private PilMedianFilter() {}
}

File: hash/swift/RosettaSquintHash/Sources/RosettaSquintHash/Internal/PilMedianFilter.swift
import Foundation

/// PIL-compatible MedianFilter(size=3) on UInt8 grayscale (row-major).
///
/// 3×3 windowed median with edge-replication (clamp) boundary.
/// For each pixel, gathers the 9 values in the window (clamping out-of-bounds
/// coordinates to the nearest valid pixel), sorts them, and returns index 4.
/// This matches Pillow's `ImageFilter.MedianFilter()` with size=3.

func pilMedianFilter(_ src: [UInt8], width: Int, height: Int) -> [UInt8] {
    var result = [UInt8](repeating: 0, count: width * height)

    for y in 0..<height {
        for x in 0..<width {
            var window = [UInt8](repeating: 0, count: 9)
            var idx = 0
            for dy in -1...1 {
                for dx in -1...1 {
                    let yy = max(0, min(height - 1, y + dy))
                    let xx = max(0, min(width  - 1, x + dx))
                    window[idx] = src[yy * width + xx]
                    idx += 1
                }
            }
            // Partial insertion sort to find median (index 4 of 9)
            window.sort()
            result[y * width + x] = window[4]
        }
    }

    return result
}
